libraries/ms-drivers/test/test_mcp3427_adc.c

```diff
@@ -0,0 +1,371 @@
+#include "delay.h"
+#include "event_queue.h"
+#include "i2c.h"
+#include "interrupt.h"
+#include "log.h"
+#include "mcp3427_adc.h"
+#include "mcp3427_adc_defs.h"
+#include "ms_test_helpers.h"
+#include "soft_timer.h"
+#include "test_helpers.h"
+#include "unity.h"
+
+#define TEST_ADDR_PIN_0 MCP3427_PIN_STATE_LOW
+#define TEST_ADDR_PIN_1 MCP3427_PIN_STATE_LOW
+#define TEST_I2C_PORT I2C_PORT_2
+
+// these are used for a second address in |test_multiple_mcp3427s|
+#define TEST_SECONDARY_ADDR_PIN_0 MCP3427_PIN_STATE_HIGH
+#define TEST_SECONDARY_ADDR_PIN_1 MCP3427_PIN_STATE_HIGH
+#define TEST_SECONDARY_I2C_PORT I2C_PORT_2
+
+#define TEST_AMP_GAIN MCP3427_AMP_GAIN_1
+
+#define TEST_CONFIG_PIN_I2C_SCL \
+  { GPIO_PORT_B, 10 }
+#define TEST_CONFIG_PIN_I2C_SDA \
+  { GPIO_PORT_B, 11 }
+
+typedef enum {
+  TEST_MCP3427_DATA_TRIGGER_EVENT = 0,
+  TEST_MCP3427_DATA_READY_EVENT,
+  NUM_MCP3427_TEST_EVENTS,
+} TestMcp3427Event;
+
+// Like MS_TEST_HELPERS_ASSERT_EVENT, but only checks ID and not data
+#define TEST_ASSERT_EVENT_WITH_ID(e, event_id) \
+  ({                                           \
+    TEST_ASSERT_OK(event_process(&(e)));       \
+    TEST_ASSERT_EQUAL((event_id), (e).id);     \
+  })
+
+static uint16_t s_times_callback_called = 0;
+static void *s_callback_context = NULL;
+static uint16_t s_times_fault_callback_called = 0;
+static void *s_fault_callback_context = NULL;
+
+static void prv_callback(int16_t value_ch1, int16_t value_ch2, void *context) {
+  LOG_DEBUG("Callback called: value_ch1=%d, value_ch2=%d\n", value_ch1, value_ch2);
+  s_times_callback_called++;
+  s_callback_context = context;
+}
+
+static void prv_fault_callback(void *context) {
+  LOG_WARN("Fault callback called!");
+  s_times_fault_callback_called++;
+  s_fault_callback_context = context;
+}
+
+void setup_test(void) {
+  event_queue_init();
+  interrupt_init();
+  soft_timer_init();
+
+  I2CSettings i2c_settings = {
+    .speed = I2C_SPEED_FAST,
+    .sda = TEST_CONFIG_PIN_I2C_SDA,
+    .scl = TEST_CONFIG_PIN_I2C_SCL,
+  };
+  i2c_init(TEST_I2C_PORT, &i2c_settings);
+
+  s_times_callback_called = s_times_fault_callback_called = 0;
+  s_callback_context = s_fault_callback_context = NULL;
```

Comment:
I'm faaaaairly certain this breaks google C++ style guide but can't prove it with a skim through. Mostly just a nitpick though, whatever really.